File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/composers/TableComposer.java
package com.wwdablu.soumya.simplypdf.composers;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.wwdablu.soumya.simplypdf.SimplyPdfDocument;
import com.wwdablu.soumya.simplypdf.composers.models.TableProperties;
import com.wwdablu.soumya.simplypdf.composers.models.TextProperties;
import com.wwdablu.soumya.simplypdf.composers.models.cell.Cell;
import com.wwdablu.soumya.simplypdf.composers.models.cell.TextCell;

import java.util.List;

public class TableComposer extends GroupComposer {

    private Paint borderPainter;
    private TableProperties colProperties;

    private TextComposer textComposer;

    public TableComposer(@NonNull SimplyPdfDocument simplyPdfDocument) {

        this.simplyPdfDocument = simplyPdfDocument;
        this.borderPainter = new Paint(Paint.ANTI_ALIAS_FLAG);
        initProperties();
        initSupportedComposers(simplyPdfDocument);
    }

    public void draw(@NonNull List<List<Cell>> cellList) {

        if(cellList.size() == 0) {
            return;
        }

        int largestHeight = 0;
        for(List<Cell> rowCellList : cellList) {

            for(Cell cell : rowCellList) {

                int cellHeight = 0;
                if(cell instanceof TextCell) {
                    cellHeight = textComposer
                        .write(((TextCell) cell).text, ((TextCell) cell).properties,
                            cell.width, cell.verticalPadding, true, 0, cell.horizontalPadding, false);
                }

                if(largestHeight < cellHeight) {
                    largestHeight = cellHeight;
                }
            }

            if(!canFitContentInPage(largestHeight)) {
                simplyPdfDocument.newPage();
            }

            int bitmapXTranslate = simplyPdfDocument.getLeftMargin();
            int arrayLength = rowCellList.size();
            for(int rowIndex = 0; rowIndex < arrayLength; rowIndex++) {

                Cell cellData = rowCellList.get(rowIndex);
                if(cellData instanceof TextCell) {
                    textComposer.write(((TextCell) cellData).text, ((TextCell) cellData).properties,
                        cellData.width, cellData.verticalPadding,
                            true, bitmapXTranslate, cellData.horizontalPadding, true);
                    bitmapXTranslate += (cellData.width);
                }
            }

            simplyPdfDocument.addContentHeight(largestHeight);
            drawBorders(getPageCanvas(), largestHeight, rowCellList);
            largestHeight = 0;
        }
    }

    public void setProperties(@Nullable TableProperties properties) {

        if(properties != null) {
            this.colProperties = properties;
        } else {
            initProperties();
        }
    }

    @Override
    void clean() {
        super.clean();
        borderPainter = null;
    }

    @Override
    public String getComposerName() {
        return TableComposer.class.getName();
    }

    private void initProperties() {
        this.colProperties = new TableProperties();
        colProperties.borderColor = "#000000";
        colProperties.borderWidth = 1;
    }

    private void initSupportedComposers(@NonNull SimplyPdfDocument simplyPdfDocument) {
        textComposer = new TextComposer(simplyPdfDocument);
    }

    private void drawBorders(@NonNull Canvas canvas, int maxHeight, List<Cell> rowCellList) {

        canvas.save();
        canvas.translate(simplyPdfDocument.getLeftMargin(), simplyPdfDocument.getPageContentHeight() - maxHeight);
        borderPainter.setColor(Color.parseColor(colProperties.borderColor));

        int colIndex = 0;
        for(Cell cell : rowCellList) {

            //Left border
            if(colIndex == 0) {
                canvas.drawLine(0, 0, 0, maxHeight, borderPainter);
            }

            //Top border
            canvas.drawLine(0, 0, cell.width, 0, borderPainter);

            //Right border
            canvas.drawLine(cell.width, 0, cell.width, maxHeight, borderPainter);

            //Bottom border
            canvas.drawLine(cell.width, maxHeight, 0, maxHeight, borderPainter);

            ++colIndex;
            canvas.translate(cell.width, 0);
        }

        canvas.restore();
    }
}

File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/composers/TableComposer.kt
package com.wwdablu.soumya.simplypdf.composers

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import com.wwdablu.soumya.simplypdf.SimplyPdfDocument
import com.wwdablu.soumya.simplypdf.composers.properties.TableProperties
import com.wwdablu.soumya.simplypdf.composers.properties.cell.Cell

class TableComposer(simplyPdfDocument: SimplyPdfDocument) : GroupComposer(simplyPdfDocument) {

    private val borderPainter = Paint(Paint.ANTI_ALIAS_FLAG)

    override fun getTypeHandler(): String = "table"

    fun draw(cellList: List<List<Cell>>, properties: TableProperties) {

        if (cellList.isEmpty()) {
            return
        }

        var largestHeight: Int
        var rowIndex = 0
        val rowCount = cellList.size
        var alignTranslate = 0

        //If center aligned is set, then render it
        if(properties.align == TableProperties.ALIGN_CENTER) {
            val maxWidth = getMaxWidth(cellList)

            if (maxWidth < simplyPdfDocument.usablePageWidth) {
                alignTranslate = (simplyPdfDocument.usablePageWidth - maxWidth) / 2
            }
        }

        cellList.forEach { rowCellList ->

            largestHeight = getLargestCellSize(rowCellList)

            if (!canFitContentInPage(largestHeight)) {
                simplyPdfDocument.newPage()
                simplyPdfDocument.insertEmptyLines(1)
            }

            var bitmapXTranslate = simplyPdfDocument.startMargin + alignTranslate
            rowCellList.forEach { cell ->
                cell.apply {
                    render(bitmapXTranslate)
                    bitmapXTranslate += getCellWidth()
                }
            }

            if(properties.drawBorder) {
                drawBorders(pageCanvas, largestHeight.toFloat(), alignTranslate,
                    rowIndex, rowCount, rowCellList, properties)
            }

            rowIndex++
            simplyPdfDocument.addContentHeight(largestHeight)
        }
    }

    private fun getLargestCellSize(rowCellList: List<Cell>) : Int {

        var largestHeight = 0
        rowCellList.forEach {
            it.setDocument(simplyPdfDocument)
            largestHeight = largestHeight.coerceAtLeast(it.getCellHeight())
        }
        return largestHeight
    }

    private fun drawBorders(canvas: Canvas,
                            maxHeight: Float,
                            xTranslate: Int,
                            rowIndex: Int,
                            rowCount: Int,
                            rowCellList: List<Cell>,
                            tableProperties: TableProperties) {

        borderPainter.color = Color.parseColor(tableProperties.borderColor)

        canvas.save()
        canvas.translate(simplyPdfDocument.startMargin.toFloat() + xTranslate,
            simplyPdfDocument.pageContentHeight.toFloat())

        var colIndex: Int = 0

        rowCellList.forEach { cell ->
            //Left border
            if (colIndex == 0 && canDrawLeftBorder(tableProperties, rowIndex, rowCount, colIndex, rowCellList.size)) {
                canvas.drawLine(0f, 0f, 0f, maxHeight, borderPainter)
            }

            //Top border
            if(canDrawTopBorder(tableProperties, rowIndex, rowCount, colIndex, rowCellList.size)) {
                canvas.drawLine(
                    0f, 0f, cell.getCellWidth().toFloat(),
                    0f, borderPainter
                )
            }

            //Right border
            if(canDrawRightBorder(tableProperties, rowIndex, rowCount, colIndex, rowCellList.size)) {
                canvas.drawLine(
                    cell.getCellWidth().toFloat(), 0f,
                    cell.getCellWidth().toFloat(), maxHeight, borderPainter
                )
            }

            //Bottom border
            if(canDrawBottomBorder(tableProperties, rowIndex, rowCount, colIndex, rowCellList.size)) {
                canvas.drawLine(
                    cell.getCellWidth().toFloat(), maxHeight,
                    0f, maxHeight, borderPainter
                )
            }

            colIndex++
            canvas.translate(cell.getCellWidth().toFloat(), 0f)
        }

        canvas.restore()
    }

    private fun canDrawLeftBorder(prop: TableProperties,
                                  rowIndex: Int,
                                  rowCount: Int,
                                  colIndex: Int,
                                  colCount: Int) : Boolean {

        if(prop.borderStyle or TableProperties.BORDER_ALL == TableProperties.BORDER_ALL ||
            prop.borderStyle or TableProperties.BORDER_VERTICAL == TableProperties.BORDER_VERTICAL) {
            return true
        }

        if(prop.borderStyle or TableProperties.BORDER_INNER == TableProperties.BORDER_INNER && colIndex != 0) {
            return true
        }

        return prop.borderStyle or TableProperties.BORDER_OUTER == TableProperties.BORDER_OUTER &&
                colIndex == 0
    }

    private fun canDrawRightBorder(prop: TableProperties,
                                  rowIndex: Int,
                                  rowCount: Int,
                                  colIndex: Int,
                                  colCount: Int) : Boolean {

        if(prop.borderStyle or TableProperties.BORDER_ALL == TableProperties.BORDER_ALL ||
            prop.borderStyle or TableProperties.BORDER_VERTICAL == TableProperties.BORDER_VERTICAL) {
            return true
        }

        if(prop.borderStyle or TableProperties.BORDER_INNER == TableProperties.BORDER_INNER &&
            colIndex != colCount - 1) {
            return true
        }

        return prop.borderStyle or TableProperties.BORDER_OUTER == TableProperties.BORDER_OUTER &&
                colIndex == colCount - 1
    }

    private fun canDrawTopBorder(prop: TableProperties,
                                 rowIndex: Int,
                                 rowCount: Int,
                                 colIndex: Int,
                                 colCount: Int) : Boolean {

        if(prop.borderStyle or TableProperties.BORDER_ALL == TableProperties.BORDER_ALL ||
            prop.borderStyle or TableProperties.BORDER_HORIZONTAL == TableProperties.BORDER_HORIZONTAL) {
            return true
        }

        if(prop.borderStyle or TableProperties.BORDER_INNER == TableProperties.BORDER_INNER && rowIndex != 0) {
            return true
        }

        return prop.borderStyle or TableProperties.BORDER_OUTER == TableProperties.BORDER_OUTER &&
                rowIndex == 0
    }

    private fun canDrawBottomBorder(prop: TableProperties,
                                 rowIndex: Int,
                                 rowCount: Int,
                                 colIndex: Int,
                                 colCount: Int) : Boolean {

        if(prop.borderStyle or TableProperties.BORDER_ALL == TableProperties.BORDER_ALL ||
            prop.borderStyle or TableProperties.BORDER_HORIZONTAL == TableProperties.BORDER_HORIZONTAL) {
            return true
        }

        if(prop.borderStyle or TableProperties.BORDER_INNER == TableProperties.BORDER_INNER &&
            rowIndex != rowCount - 1) {
            return true
        }

        return prop.borderStyle or TableProperties.BORDER_OUTER == TableProperties.BORDER_OUTER &&
                rowIndex == rowCount - 1
    }

    private fun getMaxWidth(cellList: List<List<Cell>>) : Int {

        var maxWidth = 0

        cellList.forEach { row ->

            var width = 0
            row.forEach { item ->
                width += item.width
            }

            if(width > maxWidth) {
                maxWidth = width
            }
        }

        return maxWidth
    }
}